File: 2023/21.py
input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n")
start = [0, 0]
for i in range(len(input)):
    for j in range(len(input[0])):
        if input[i][j] == "S":
            start = [i, j]
            break


def bfs(steps):
    if len(steps) > 64:
        return len(steps[-1])

    steps.append([])
    for c in steps[-2]:
        for dir in [[0, 1], [0, -1], [1, 0], [-1, 0]]:
            new_c = [c[0] + dir[0], c[1] + dir[1]]
            if (
                new_c[0] < 0
                or new_c[0] >= len(input)
                or new_c[1] < 0
                or new_c[1] >= len(input[0])
                or input[new_c[0]][new_c[1]] == "#"
                or new_c in steps[-1]
            ):
                continue
            steps[-1].append(new_c)
    return bfs(steps)


print(bfs([[start]]))

steps_memo = {}
stack = [start + [0, 0, 0]]

while stack:
    x, y, map_x, map_y, steps = stack.pop(0)
    if x < 0:
        map_x -= 1
        x += len(input)
    elif x >= len(input):
        map_x += 1
        x -= len(input)
    if y < 0:
        map_y -= 1
        y += len(input[0])
    elif y >= len(input[0]):
        map_y += 1
        y -= len(input[0])

    if (
        input[x][y] == "#"
        or repr([x, y, map_x, map_y]) in steps_memo
        or map_x < -1
        or map_x > 1
        or map_y < -1
        or map_y > 1
    ):
        continue

    steps_memo[repr([x, y, map_x, map_y])] = steps

    for dir in [[0, 1], [0, -1], [1, 0], [-1, 0]]:
        stack.append([x + dir[0], y + dir[1], map_x, map_y, steps + 1])


memo = {}


def repeat(steps, map_corner):
    total_steps = 26_501_365
    if repr([steps, map_corner, total_steps]) in memo:
        return memo[repr([steps, map_corner, total_steps])]

    res = 0
    for i in range(1, (total_steps - steps) // len(input) + 1):
        if steps + len(input) * i > total_steps or (steps + len(input) * i) % 2 == 0:
            continue

        res += 1
        if not map_corner:
            res += i

    memo[repr([steps, map_corner, total_steps])] = res
    return res


res = 0
for i in range(len(input)):
    for j in range(len(input[0])):
        if not repr([i, j, 0, 0]) in steps_memo:
            continue

        for map_x in [-1, 0, 1]:
            for map_y in [-1, 0, 1]:
                steps = steps_memo[repr([i, j, map_x, map_y])]
                if steps % 2 == 1:
                    res += 1

                if map_x == 0 and map_y == 0:
                    continue

                res += repeat(steps, map_x == 0 or map_y == 0)

print(res)

File: 2023/21.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input')).strip.split("\n").map(&:chars)

start = [input.index { _1.include?('S') }, input[input.index { _1.include?('S') }].index('S')]

def bfs(input, steps)
  return steps.last.count if steps.size > 64

  steps << []
  steps[-2].each do |c|
    [[0, 1], [0, -1], [1, 0], [-1, 0]].each do |dir|
      new_c = [c[0] + dir[0], c[1] + dir[1]]
      if new_c[0] >= input.size || new_c[0] < 0 ||
         new_c[1] > input[0].size || new_c[1] < 0 ||
         input[new_c[0]][new_c[1]] == '#' ||
         steps[-1].include?(new_c)
        next
      end

      steps[-1] << new_c
    end
  end
  bfs(input, steps)
end
p bfs(input, [[start]])

steps_memo = {}
stack = [start + [0, 0, 0]]
until stack.empty?
  x, y, map_x, map_y, steps = stack.shift
  if x < 0
    map_x -= 1
    x += input.size
  elsif x >= input.size
    map_x += 1
    x -= input.size
  end
  if y < 0
    map_y -= 1
    y += input[0].size
  elsif y >= input[0].size
    map_y += 1
    y -= input[0].size
  end

  next if input[x][y] == '#' || steps_memo.key?([x, y, map_x, map_y])
  next if map_x < -1 || map_x > 1 || map_y < -1 || map_y > 1

  steps_memo[[x, y, map_x, map_y]] = steps

  [[0, 1], [0, -1], [1, 0], [-1, 0]].each do |dir|
    stack << [x + dir[0], y + dir[1], map_x, map_y, steps + 1]
  end
end

@memo = {}
def repeat(steps, map_corner, input)
  total_steps = 26_501_365
  return @memo[[steps, map_corner, total_steps]] if @memo.key?([steps, map_corner, total_steps])

  res = 0
  (1..(total_steps - steps) / input.size).each do |i|
    next unless steps + input.size * i <= total_steps && (steps + input.size * i).odd?

    res += 1
    res += i unless map_corner
  end
  @memo[[steps, map_corner, total_steps]] = res
  res
end

res = 0
(0...input.size).each do |i|
  (0...input[0].size).each do |j|
    next unless steps_memo.key?([i, j, 0, 0])

    (-1..1).each do |map_x|
      (-1..1).each do |map_y|
        steps = steps_memo[[i, j, map_x, map_y]]
        res += 1 if steps.odd?

        next if map_x == 0 && map_y == 0

        res += repeat(steps, map_x == 0 || map_y == 0, input)
      end
    end
  end
end

p res
